Json tests

catamorphism opened the pull request:
<!--
Thank you for your pull request!

* General info on contributing: please see https://github.com/unicode-org/icu/blob/main/CONTRIBUTING.md
* Ticket numbers for minor changes: for minor changes (ex: docs typos), you can reuse one of the open catch-all tickets for our next release
  - ICU 76 ticket: docs minor fixes: typos/etc./version updates / User Guide & API docs: ICU-22722
  - ICU 76 ticket: code warnings/version updates: ICU-22721
* Contributors license agreement (CLA): 
  You will be automatically asked to sign the CLA before the PR is accepted.
  To sign the CLA: https://cla-assistant.io/unicode-org/icu

  For terms of use and license, see https://www.unicode.org/terms_of_use.html
-->

##### Checklist

- [ ] Required: Issue filed: https://unicode-org.atlassian.net/browse/ICU-22746
- [ ] Required: The PR title must be prefixed with a JIRA Issue number. <!-- For example: "ICU-1234 Fix xyz" -->
- [ ] Required: The PR description must include the link to the Jira Issue, for example by completing the URL in the first checklist item
- [ ] Required: Each commit message must be prefixed with a JIRA Issue number. <!-- For example: "ICU-1234 Fix xyz" -->
- [ ] Issue accepted (done by Technical Committee after discussion)
- [ ] Tests included, if applicable
- [ ] API docs and/or User Guide docs changed or added, if applicable

catamorphism commented:
cc @echeran -- I didn't fix the copyright scan errors since you mentioned you would help with legal signoff re: licensing for the nlohmann/json library.

cc @srl295 @roubert -- I tried to follow what you suggested with vendoring the library; let me know if this is what you had in mind. I blatantly copy-pasted the pull-from-upstream.sh and UPDATING.md files from the `vendor/double-conversion` directory, but if you think it's worth generalizing the instructions and shell script, let me know.

srl295 commented:
good work and a couple of comments
- i see two copies of ujson.h and probably only need one
- does the entire repo need to be vendored since only one file is used? perhaps only the .h file and the license? 
- it might be good to have a file with the latest hash and version or such (in the vendor directory)

catamorphism commented:
> good work and a couple of comments
> 
>     * i see two copies of ujson.h and probably only need one
> 

Thanks for spotting that, fixed.

>     * does the entire repo need to be vendored since only one file is used? perhaps only the .h file and the license?
> 

Probably not, no. Changed this in [24de7c4](https://github.com/unicode-org/icu/pull/2980/commits/24de7c47a5ae2484ef09d7088449a4dad402d7c5) and removed the extra files.

>     * it might be good to have a file with the latest hash and version or such (in the vendor directory)

Do you mean a manually updated file, or changing the pull-from-upstream.sh script to update the file when it's run?

srl295 commented on vendor/json/UPDATING.md:
```diff
@@ -0,0 +1,52 @@
+
+## Next Steps
+
+Build and test ICU4C, and send the PR for review.
```
```suggestion
Build and test ICU4C, and send the PR for review.

Verify that `LICENSE.MIT` has not changed, may need Unicode legal review if it has.
```

srl295 reviewed (changes requested):
- should have some mention of checking that the license doesn't change
-

srl295 commented:
- [ ] also this PR needs to update the top level `LICENSE` file to include a copy of LICENSE.MIT, see other examples

echeran commented on vendor/json/upstream/LICENSE.MIT:
Also, indicate which files that the license to: "Mr. Lohman's copyright notice and license will have to be reproduced in the ICU license file making clear what imported content it applies to."

echeran commented:
I double checked with Google's legal counsel for open source, who gave approval continent on guidelines that already match ICU's current processes for vendoring 3rd party sources:

> It will be fine to vendor MIT-licensed code into ICU. ... The LICENSE file that lives alongside the nlohmann/json code in the `/vendor` subdirectory of ICU needs to inherit the specific MIT license file at https://github.com/nlohmann/json/blob/develop/LICENSE.MIT with its upstream copyright notice intact.

catamorphism commented:
> * should have some mention of checking that the license doesn't change

I think your suggested change, which I incorporated, addresses this?

srl295 reviewed (changes requested):
@catamorphism specifically:  please update the top level [LICENSE](https://github.com/unicode-org/icu/blob/main/LICENSE#L45) to incorporate a section identifying and close

catamorphism commented:
OK, I think I've incorporated all the licensing changes you suggested, @echeran and @srl295 -- can you look again? Thanks!

srl295 commented on .cpyskip.txt:
```diff
@@ -117,3 +118,4 @@ tools/unicodetools/*
 # vendor
 #
 vendor/double-conversion/upstream/*
+vendor/json/upstream/*
```
will it need to also skip the file in its copied in location? might be needed

catamorphism replied:
I changed the copied-in `json-json.hpp` file to include the Unicode license at the top, and the pull-from-upstream script should preserve the change, so I think that's OK? Though I'll look at the failing copyright-scan output.

srl295 replied:
it looked like that was fine, just the readme

srl295 commented on icu4c/source/test/testdata/message2/README.txt:
```diff
@@ -0,0 +1,18 @@
+
+Tests in the `spec/` subdirectory are taken from https://github.com/unicode-org/message-format-wg/blob/main/test
+and need to be manually updated if the contents change upstream.
+
```
```suggestion

Copyright © 2024 Unicode, Inc. Unicode and the Unicode Logo are registered trademarks of Unicode, Inc. in the United States and other countries.
```

(remember to accept the comment, if you want to accept it this way, with `ICU-22746` as the beginning of the commit messag)

srl295 commented:
@catamorphism all LGTM!  
- probably want to squash it per ICU prefs - https://us-central1-dev-infra-273822.cloudfunctions.net/unicode-github-bot/info/unicode-org/icu/2980  (the 'Details' under the single-commit check) will do it for you
- @echeran are you driving the ticket and the merge?

echeran reviewed (approved):
LGTM, thanks!

catamorphism commented:
Thanks! I've squashed.
